### ISSUE TITLE
What should we do with the classes once loaded?

### ISSUE THREAD

**Issue by JuanFdS** [thin-languages/slimdown#13]
We are already able to locate and load the classes to the JVM, what should we do now to get them to add functionality to the IDE?
The first idea that came to me was filtering the classes that are the main classes of each plugin and sending them a method they are required to know. This could be as simple as filtering the classes that can answer to the method main(args:Array[String]) and calling that main method.
However, I'd like this to be discussed before we start implementing it, so, what do you think?

**Comment by JuanchiRios** [comment]
Im okay with this thought. But what do you think about receiving a Try[IdeView] and return Try[IdeView] ? So once we get a list of main methods from main classes, we fold it.

And now with this idea instead of having a main method we could just inspect all methods and filter methods which returns Try[IdeView] and received only one parameter Try[IdeView]. On the one hand you don't need to name "main" a method, on the other hand the transformations may have an order and with this solutions we could screw things up. So idk maybe your solution it's better because you delegate transformation order in  plugin-developers(and i think that is okay)

**Comment by JuanFdS** [comment]
That would be better, passing the entire IDE as a parameter and each plugin modifies what it needs and returns a new instance of it.
However, I think we should require the main method to be called in some specific way so we can distinguish it from the rest, otherwise if the plugins have auxiliar methods that go from Try[IdeView] to Try[IdeView] we would end up calling them when we shouldn't.

**Comment by micaelafmartino** [comment]
I agree with having a main method in the plugins, which receives some representation of the IDE, and returns another, wrapped in a Try. That main method could be apply, so we can think about plugins as functions that transforms our IDE, and then, like @JuanchiRios said, fold the plugins list to obtain the final composition. 

I don't think we should pass only the view, because each plugin must also add/alter some backend functionality.

**Comment by JuanFdS** [comment]
I like the idea of using the apply method, but I don't know if we will be able to take advantage of the sintax sugar that it provides because we will be loading the classes via classloader so I guess we will have to use reflection to send messages to those classes' instances.
And btw, I realised how we could overcome the problem of filtering classes by the name of their methods: it would me more practical and elegant to filter the main classes by their type, so we could require those to extend from a known trait that we defined, like IdePlugin or something like that.

**Comment by nscarcella** [comment]
In my experience working with plugins, it is generally better to establish "kinds" of plugins to provide better interfaces and a more guided experience for developers.

Plugins affect pretty much everything in an IDE. The GUI is one good example, but not the only one. If only one type of plugin is available soon all plugins will have to receive a lot of arguments and the implementations will be monolithic and messy.

Think about this: A plugin may have to build a view, set preferences, provide support for different types of files, expose logic to be available for other plugins and more. If all this has to be done in one method (even using auxiliar methods) it soon becomes hard to maintain. Also, some of this tasks may require parameters (a gui builder, for example). Make all plugins receive this argument and any change to the parameter's contract will break all plugins, even the ones that doesn't use the troubled parameter.

A good approach may be to define a series of interfaces that a plugin can expose in order to be recognized as such. So, for example, instead of tracking a class because it has a main method that receives an IDEView parameter, track classes that extend GUIPlugin:

``` scala
trait GUIPlugin {
  def buildGUI(currentGUI: IDEVIew): IDEView // Plugins implements this to make stuff happen
}
```

That way, each plugin can extend one or more of this smaller contracts, scoping the objectives and parameters better and making it clearer to the developer what should be done where.

A generic Plugin supertype could be created to make the task of identifying classes easier.

Another advantage of this approach is that, once a class has reflectively been identified as a plugin, you have a type you can cast it to, so you don't have to keep using reflective calls.

A fancier way, similar to what the latest versions of Eclipse do, is treat parameters of some type as "hooks". So you define classes with any method you want, but add certain parameters to the signatures. This parameters are used as booth a dependency to do something as well as a declaration that the plugin wants to take part of that task. The IDE defines some kind of life cicle and calls the plugins methods whenever a change occurs for that kind of parameter.

So, for example, you can define a type GUIBuilder that exposes all that is needed to customize the GUI, and decide that you will call all methods that expect that parameter when the GUI needs to be updated (this can happen at the start of the IDE or in some other time too).

``` scala
class MySpecificPlugin {
  def doSomethingIWant(guiBuilder: Option[GUIBuilder]) {
     //some code that will be executed by the IDE whenever gui building is required
  }
}
```

This approach is a bit more interesting than the first one, since it is really flexible, and allows a more dynamic way of executing the plugins (not all the plugins need to be executed on start, making things faster) and does not require the IDE team to hard-code "plugin types". The problem is that it may be a bit harder to implement.

Here are some other ideas we could take in consideration:
- **Declarativity**
  As with so many other things, procedural definitions are tricky to follow and hard to debug. We should focus on trying to make plugins interface so each plugin have to be implemented in an as declarative as possible way. That would mean that order is not important and each plugin can be think of in an isolated way. Plugins should also not rely on side effect, but base themselves in returning values or their interaction with other objects.
- **Undoability**
  One of the most unpleasant things about plugins are long waiting times. Restart an IDE that needs to load a lot of stuff takes time, so clean restart is something we could try to avoid as much as possible. A great feature some IDEs have is the ability to "unplug" a plugin; turn it on/off while running, sort of speak.
- **Testability**
  Life is always a bit harder when you have to extend a piece of software somebody else developed. Booth the IDE and plugins should be as easy to test as possible, to encourage good practices and allow unexperienced developers to learn how things are done. This may be trickier than it sounds, since it is easy to loose track of what is complex while developing a tool this big, but at the very least we should make all we can so plugins can be tested in booth an isolated way and with a minimum controlled set of base plugins. Also take the time to provide mocks and matchers that simplify this task, and avoid defining interfaces that are too coupled to a certain implementation or tool.

Well, sorry to bumble around, just wanted to left this written somewhere. :)

**Comment by JuanchiRios** [comment]
So in the fancier solution we will search each class and looking for methods that have well known TypesParemeters? We could start making the lookup method defining a couple of well known classes and execute those methods

**Comment by nscarcella** [comment]
By the way, one more thing:

I'm pretty biased because I got most experience working on Java based frameworks, but since we are using Scala here it may be nice to try and make our interfaces as Scala friendly as possible.

One example of this is, instead of requiring plugins to be a class with some kind of empty constructor we could accept objects instead. Think our interfaces to use Option, Try and Lists is another good practice. We may even think about plugins as functions. I don't really have any concrete idea of where and how, but I thought to mention this.
